File: C/algorithms/searching/binary_search.c
#include <stdio.h>

int binarySearch (int arr[], int key, int low, int high) {
    int mid;
    if (low > high) return (-1); /* not found */
    mid = (low + high) / 2;
    if (arr[mid] == key)   return (mid);
    if (arr[mid] > key)
        return (binarySearch(arr, key, low, mid-1));
    else
        return (binarySearch(arr, key, mid+1, high));
}

int main () {
    int arr[] = {1, 4, 5, 6, 8, 10, 12};
    printf ("%d\n", binarySearch(arr, 6, 0, 6));
    return 0;
}

File: ADS/search/binary_search.py
#!/usr/bin/env python3

def binary_search(arr, item):
    start = 0
    end = len(arr) - 1

    while start <= end:
        mid = (start + end) // 2
        curr = arr[mid]

        if curr == item:
            return mid
        elif item < curr:
            end = mid - 1
        else:
            start = mid + 1

    return None

print(binary_search([1, 3, 4, 8, 9, 12, 15, 18, 22, 26, 29], 12))
print(binary_search([1, 3, 4, 8, 9, 12, 15, 18, 22, 26, 29], 8))
print(binary_search([1, 3, 4, 8, 9, 12, 15, 18, 22, 26, 29], 22))
